examples/utils.py

```diff
@@ -1,8 +1,61 @@
 # This module defines a set of common functions which are used in other examples.
 import time
 import datetime
-from tabulate import tabulate
+import os
+import importlib.util
+
+
+# IMPORTS
+def ensure_installed(*names):
+    """
+    Ensures a package is installed on the system, by installing it if it does not exist currently.
+
+    Args:
+        names (str): the names of the package to be checked (e.g. pandas, numpy, etc)
+
+    Returns:
+        None
+    """
+    for name in names:
+        if importlib.util.find_spec(name) is None:
+            install_package(name)
+
+def install_package(name, version=None):
+    """
+    Installs a package via Pip. If a version is supplied, will attempt to install that specific version.
+    If one is not supplied, requirements.txt will be searched to find a version.
+    If a version is still not found, the latest version of the package will be installed.
+
+    Args:
+        name (str): the name of the module (e.g. pandas, numpy, etc)
+        version (str): the specific version (if any) to import (e.g. 0.1.5, 1.0.0, etc).
 
+    Returns:
+        None
+    """
+    # Check requiements.txt for current version, if one wasn't supplied
+    if version is None:
+        reqs_file = os.path.realpath(os.path.join(__file__, "../../requirements.txt"))
+        with open(reqs_file, "r") as reqs:
+            for line in reqs:
+                if name in line:
+                    version = line.strip().split("==")[1]
+
+    # Add version if one was found or specified
+    if version is not None:
+        pip_name = f"{name}=={version}"
+    else:
+        pip_name = name
+
+    # Install the modules
+    import sys, subprocess
+    subprocess.call([sys.executable, "-m", "pip", "install", pip_name])
```

Comment:
What I had in mind originally was similar to https://jakevdp.github.io/blog/2017/12/05/installing-python-packages-from-jupyter/#How-to-use-Pip-from-the-Jupyter-Notebook, for using the jupyter shell `!<command>`. Is our outcome here any different or is it equivalent?

Reply:
That was the original approach, but it turned out to create garbage on the respective lines in the .py files generated - that's why the previous iteration had that large regex in config.py. It would also need to reside in the import section of each notebook.

Since the ! escapes to the terminal anyway, the approaches should be equivalent whether we do !{executable} or subprocess.call('executable'). When I was testing these, the notebooks were able to install and use missing modules using the current approach